### PR TITLE
OCPBUGS-46577: Provision for user-configured endpoints for IBMCLOUD_IAM_API_ENDPOINT, IBMCLOUD_IAM_API_ENDPOINT,etc

### PR DESCRIPTION
**Pull request by kishen-v**
This PR adds changes to add flexibility of configuring custom endpoints for the CSI Driver, such as `IBMCLOUD_IAM_API_ENDPOINT`,`IBMCLOUD_RESOURCE_CONTROLLER_ENDPOINT`,`IBMCLOUD_POWER_API_ENDPOINT`, etc. which are mapped to `API`,`ResourceController` and `Power` respectively.

This allows the CSI driver to be configured to use a non default endpoint when deployed from the `ibm-powervs-block-csi-driver-operator`. 

For example:

```
    serviceEndpoints:
    - name: API
      url: "https://iam.ibmcloud-test.com"
```

References: 
https://github.com/Karthik-K-N/machine-api-provider-powervs/commit/1064d35a044fdae1cb00d105878f47ae91fc15f7
https://github.com/openshift/machine-api-provider-powervs/pull/76

### CODE REVIEW

**Comment by kishen-v** [comment]
cc: @Karthik-K-N

File: pkg/operator/starter.go [hunk 1]
```diff
@@ -209,3 +218,48 @@ func extractOperatorStatus(obj *unstructured.Unstructured, fieldManager string)
...
+		if infra.Status.PlatformStatus == nil || infra.Status.PlatformStatus.PowerVS == nil {
+			return nil
+		}
+		serviceEndPoints := infra.Status.PlatformStatus.PowerVS.ServiceEndpoints
+		if len(serviceEndPoints) == 0 {
+			return nil
+		}
+		var containerEnvVars []corev1.EnvVar
+		for _, serviceEndPoint := range serviceEndPoints {
+			containerEnvVars = append(containerEnvVars, corev1.EnvVar{
```
Comment (Karthik-K-N) [inline pkg/operator/starter.go:237]: We are interested only in 3 endpoints, The service endpoint may contain several other, So I think we should set only these 3.

File: pkg/operator/starter_test.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+									{
+										Name:  "IBMCLOUD_RESOURCE_CONTROLLER_API_ENDPOINT",
+										Value: "https://rc.ibmcloud-test.com",
+									},
+									{
+										Name:  "IBMCLOUD_POWER_API_ENDPOINT",
+										Value: "https://pi.ibmcloud-test.com",
+									},
+									{
+										Name:  "IBMCLOUD_COS_ENDPOINT",
```
Comment (Karthik-K-N) [inline pkg/operator/starter_test.go:187]: Does this endpoint matter for CSI driver?

Comment (kishen-v) [reply pkg/operator/starter_test.go:187]: I don't think apart from `IBMCLOUD_RESOURCE_CONTROLLER_API_ENDPOINT`, `IBMCLOUD_POWER_API_ENDPOINT` and `IBMCLOUD_IAM_API_ENDPOINT`, anything else would be optional. The driver, at the moment depends only on the above three. 

Should we consider limiting the scope?

File: pkg/operator/starter.go (hunk 1 above)
Comment (kishen-v) [reply pkg/operator/starter.go:237]: Sure.. I'll have the changes made.

File: pkg/operator/starter_test.go (hunk 2 above)
Comment (kishen-v) [reply pkg/operator/starter_test.go:187]: Will reduce the scopw as discussed. Thanks!

File: pkg/operator/starter_test.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+					URL:  "https://iam.ibmcloud-test.com",
+				},
+				{
+					Name: "rc",
+					URL:  "https://rc.ibmcloud-test.com",
+				},
+				{
+					Name: "pi",
+					URL:  "https://pi.ibmcloud-test.com",
+				},
```
[the remark this answers is not in the document]
Comment (kishen-v) [reply pkg/operator/starter_test.go:77]: Agree, added a test-case which includes a dummy field.

File: pkg/operator/starter.go [hunk 4]
```diff
@@ -209,3 +218,36 @@ func extractOperatorStatus(obj *unstructured.Unstructured, fieldManager string)
...
+				Name:  endPointKeyToEnvNameMap[serviceEndPoint.Name],
```
[the remark this answers is not in the document]
Comment (kishen-v) [reply pkg/operator/starter.go:238]: Agree. I have addressed this and have added a suitable testcase.

File: pkg/operator/starter.go [hunk 5]
```diff
@@ -35,6 +39,15 @@ const (
 	cloudCredSecretName   = "ibm-powervs-block-cloud-credentials"
 	metricsCertSecretName = "ibm-powervs-block-csi-driver-controller-metrics-serving-cert"
 	trustedCAConfigMap    = "ibm-powervs-block-csi-driver-trusted-ca-bundle"
+	infrastructureName    = "cluster"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (kishen-v) [reply pkg/operator/starter.go:45]: Sure. I have added documentation.

File: README.md [hunk 6]
```diff
@@ -17,6 +17,20 @@ oc scale --replicas=0 deploy/cluster-storage-operator -n openshift-cluster-stora
...
+   serviceEndpoints:
+    - name: rc
+      url: https://resource-controller.test.cloud.ibm.com
+```
+Reference:
+```shell
+iam - IBMCLOUD_IAM_API_ENDPOINT
+rc - IBMCLOUD_RESOURCE_CONTROLLER_API_ENDPOINT
+pi - IBMCLOUD_POWER_API_ENDPOINT
+
```
[the remark this answers is not in the document]
Comment (kishen-v) [reply README.md:32]: Addressed!

File: README.md [hunk 7]
```diff
@@ -17,6 +17,20 @@ oc scale --replicas=0 deploy/cluster-storage-operator -n openshift-cluster-stora
 oc -n openshift-cluster-csi-drivers delete deployment.apps/powervs-block-csi-driver-operator deployment.apps/ibm-powervs-block-csi-driver-controller daemonset.apps/ibm-powervs-block-csi-driver-node
 ```
 
+# Define custom endpoints for the CSI driver:
+Include a `serviceEndpoints` section to configure non-default endpoints. For example:
```
[the remark this answers is not in the document]
Comment (kishen-v) [reply README.md:21]: I've hyper linked the documentation to serviceEndpoints. Hope that's okay.

File: pkg/operator/starter.go [hunk 8]
```diff
@@ -161,11 +177,6 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
 		assets.ReadFile,
 		"servicemonitor.yaml",
 	)
-
-	if err != nil {
-		return err
-	}
-
```
Comment (kishen-v) [inline pkg/operator/starter.go:168]: Redundant from :
https://github.com/openshift/ibm-powervs-block-csi-driver-operator/pull/79/files#diff-0d623dfd885adb20f991bda4c2453aebd732ca6dbb4d1d4be6e79805c3b48de6R85-R87

File: README.md [hunk 9]
```diff
@@ -17,6 +17,21 @@ oc scale --replicas=0 deploy/cluster-storage-operator -n openshift-cluster-stora
 oc -n openshift-cluster-csi-drivers delete deployment.apps/powervs-block-csi-driver-operator deployment.apps/ibm-powervs-block-csi-driver-controller daemonset.apps/ibm-powervs-block-csi-driver-node
 ```
 
+# Define custom endpoints for the CSI driver:
+Include a [serviceEndpoints](https://docs.openshift.com/container-platform/4.17/rest_api/config_apis/infrastructure-config-openshift-io-v1.html#status-platformstatus-powervs-serviceendpoints-2) section to configure non-default endpoints. For example:
```
Comment (Karthik-K-N) [inline README.md:21]: User wont be able to directly update the status subresoruce, They need to add it in spec we have a controlller which does some basic validation and mirros it across the status.

So you should change this link to point to spec: https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L1717-L1724

File: pkg/operator/starter.go [hunk 10]
```diff
@@ -209,3 +220,40 @@ func extractOperatorStatus(obj *unstructured.Unstructured, fieldManager string)
...
+			}
+			containerEnvVars = append(containerEnvVars, corev1.EnvVar{
+				Name:  endPointKeyToEnvNameMap[serviceEndPoint.Name],
+				Value: serviceEndPoint.URL,
+			})
+		}
+
+		for i := range deployment.Spec.Template.Spec.Containers {
+			container := &deployment.Spec.Template.Spec.Containers[i]
+			if container.Name != "csi-driver" {
```
Comment (Karthik-K-N) [inline pkg/operator/starter.go:251]: See if required this can be changed into constant.

File: README.md (hunk 9 above)
Comment (kishen-v) [reply README.md:21]: Hey, thanks for pointing out. 
I have updated the URL to : https://docs.openshift.com/container-platform/4.17/rest_api/config_apis/infrastructure-config-openshift-io-v1.html#spec-platformspec-powervs-serviceendpoints-2 which does cover the `spec`. 
I'll also all the reference to : https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L1717-L1724

File: pkg/operator/starter.go (hunk 10 above)
Comment (kishen-v) [reply pkg/operator/starter.go]: Addressed!

File: pkg/operator/starter.go [hunk 11]
```diff
@@ -209,3 +223,40 @@ func extractOperatorStatus(obj *unstructured.Unstructured, fieldManager string)
...
+			container := &deployment.Spec.Template.Spec.Containers[i]
+			if container.Name != csiDriver {
+				continue
+			}
+			container.Env = append(container.Env, containerEnvVars...)
+			return nil
+		}
+		return nil
+	}
+}
```
Comment (miyamotoh) [inline pkg/operator/starter.go:262]: @kishen-v I tested your changes for `deployment/ibm-powervs-block-csi-driver-controller`, but found that we would need to do the same for `daemonset/ibm-powervs-block-csi-driver-node` as well. So, I tried this (very similarly named) func that deals with `DaemonSet` and called from line 174 (inside `.WithCSIDriverNodeService` block);
```
+func withCustomEndpoint(infraLister v1.InfrastructureLister) csidrivernodeservicecontroller.DaemonSetHookFunc {
+       return func(_ *opv1.OperatorSpec, daemonset *appsv1.DaemonSet) error {
+               infra, err := infraLister.Get(infrastructureName)
+               if err != nil {
+                       return err
+               }
+               if infra.Status.PlatformStatus == nil || infra.Status.PlatformStatus.PowerVS == nil {
+                       return nil
+               }
+               serviceEndPoints := infra.Status.PlatformStatus.PowerVS.ServiceEndpoints
+               if len(serviceEndPoints) == 0 {
+                       return nil
+               }
+               var containerEnvVars []corev1.EnvVar
+               for _, serviceEndPoint := range serviceEndPoints {
+                       if _, ok := endPointKeyToEnvNameMap[serviceEndPoint.Name]; !ok {
+                               klog.Infof("Ignoring %q serviceEndpoint", serviceEndPoint.Name)
+                               continue
+                       }
+                       containerEnvVars = append(containerEnvVars, corev1.EnvVar{
+                               Name:  endPointKeyToEnvNameMap[serviceEndPoint.Name],
+                               Value: serviceEndPoint.URL,
+                       })
+               }
+
+               for i := range daemonset.Spec.Template.Spec.Containers {
+                       container := &daemonset.Spec.Template.Spec.Containers[i]
+                       if container.Name != csiDriver {
+                               continue
+                       }
+                       container.Env = append(container.Env, containerEnvVars...)
+                       return nil
+               }
+               return nil
+       }
+}
```
Does it make sense? If so, please incorporate this into your PR. After running this operator locally and scaling back up `cluster-storage-operator`, the `storage` CO becomes `Available`.

**Review by miyamotoh — COMMENTED** [review]
Perhaps additional code to handle `DaemonSet` needed. PTAL.

File: pkg/operator/starter.go (hunk 11 above)
Comment (kishen-v) [reply pkg/operator/starter.go:263]: Thanks @miyamotoh. I have made the changes as suggested.

**Comment by kishen-v** [comment]
cc @yussufsh @Karthik-K-N.

**Comment by yussufsh** [comment]
> cc @yussufsh @Karthik-K-N.

@kishen-v if ready for a review then please remove `WIP` from the title, and assign using the command `/cc`

File: README.md [hunk 12]
```diff
@@ -17,6 +17,24 @@ oc scale --replicas=0 deploy/cluster-storage-operator -n openshift-cluster-stora
 oc -n openshift-cluster-csi-drivers delete deployment.apps/powervs-block-csi-driver-operator deployment.apps/ibm-powervs-block-csi-driver-controller daemonset.apps/ibm-powervs-block-csi-driver-node
 ```
 
+# Define custom endpoints for the CSI driver:
+Include a [serviceEndpoints](https://docs.openshift.com/container-platform/4.17/rest_api/config_apis/infrastructure-config-openshift-io-v1.html#spec-platformspec-powervs-serviceendpoints-2) section to configure non-default endpoints, which is defined
```
Comment (yussufsh) [inline README.md:21]: To keep the link updated to the latest.
```suggestion
Include a [serviceEndpoints](https://docs.openshift.com/container-platform/latest/rest_api/config_apis/infrastructure-config-openshift-io-v1.html#spec-platformspec-powervs-serviceendpoints-2) section to configure non-default endpoints, which is defined
```

**Review by yussufsh — CHANGES_REQUESTED** [review]
Overall looks good. I will defer the approval to @Karthik-K-N

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm

File: README.md (hunk 12 above)
Comment (kishen-v) [reply README.md]: Addressed! thanks!

**Review by yussufsh — APPROVED** [review]

**Comment by mjturek** [comment]
/retitle OCPBUGS-46577: Provision for user-configured endpoints for IBMCLOUD_IAM_API_ENDPOINT, IBMCLOUD_IAM_API_ENDPOINT,etc

**Pull request merged**